UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,732 @@
+  CHAR8                               *GmaStr;
+
+  Fdt               = FdtBase;
+  Depth             = 0;
+  FrameBufferBase   = 0;
+  FrameBufferSize   = 0;
+  MinimalNeededSize = FixedPcdGet32 (PcdSystemMemoryUefiRegionSize);
+  IsHobConstructed  = FALSE;
+  NewHobList        = 0;
+  PlatformAcpiTable = NULL;
+  SmbiosTable       = NULL;
+  PciRootBridgeInfo = NULL;
+  RootBridgeCount   = 0;
+  index             = 0;
+  HobDataSize       = 0;
+  PciEnumDone       = 0;
+  BootMode          = 0;
+  MemType           = 0;
+  GmaID             = 0;
+
+  DEBUG ((DEBUG_INFO, "FDT = 0x%x  %x\n", Fdt, Fdt32ToCpu (*((UINT32 *)Fdt))));
+  DEBUG ((DEBUG_INFO, "Start parsing DTB data\n"));
+  DEBUG ((DEBUG_INFO, "MinimalNeededSize :%x\n", MinimalNeededSize));
+
+  for (Node = FdtNextNode (Fdt, 0, &Depth); Node >= 0; Node = FdtNextNode (Fdt, Node, &Depth)) {
+    NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + Node + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+    DEBUG ((DEBUG_INFO, "\n   Node(%08x)  %a   Depth %x", Node, NodePtr->Name, Depth));
+    // memory node
+    if (AsciiStrnCmp (NodePtr->Name, "memory@", AsciiStrLen ("memory@")) == 0) {
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+        if (AsciiStrCmp (TempStr, "reg") == 0) {
+          Data64        = (UINT64 *)(PropertyPtr->Data);
+          StartAddress  = Fdt64ToCpu (*Data64);
+          NumberOfBytes = Fdt64ToCpu (*(Data64 + 1));
+          DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          DEBUG ((DEBUG_INFO, "  %016lX  %016lX", StartAddress, NumberOfBytes));
+          if (!IsHobConstructed) {
+            if ((NumberOfBytes > MinimalNeededSize) && (StartAddress < BASE_4GB)) {
+              MemoryBottom     = StartAddress + NumberOfBytes - MinimalNeededSize;
+              FreeMemoryBottom = MemoryBottom;
+              FreeMemoryTop    = StartAddress + NumberOfBytes;
+              MemoryTop        = FreeMemoryTop;
+
+              DEBUG ((DEBUG_INFO, "MemoryBottom :0x%llx\n", MemoryBottom));
+              DEBUG ((DEBUG_INFO, "FreeMemoryBottom :0x%llx\n", FreeMemoryBottom));
+              DEBUG ((DEBUG_INFO, "FreeMemoryTop :0x%llx\n", FreeMemoryTop));
+              DEBUG ((DEBUG_INFO, "MemoryTop :0x%llx\n", MemoryTop));
+              mHobList         =  HobConstructor ((VOID *)(UINTN)MemoryBottom, (VOID *)(UINTN)MemoryTop, (VOID *)(UINTN)FreeMemoryBottom, (VOID *)(UINTN)FreeMemoryTop);
+              IsHobConstructed = TRUE;
+              NewHobList       = (UINTN)mHobList;
+              break;
+            }
+          }
+        }
+      }
+    } // end of memory node
+    else if (AsciiStrnCmp (NodePtr->Name, "pci-rb", AsciiStrLen ("pci-rb")) == 0) {

```

Comment:
UPL community recommends that we use "compatible" string instead of "node name" to indicate root bridge.
please help to update this parser code as well as the sample FDT table we are using.

UPL spec pci node section having this information:
https://universalpayload.github.io/spec/chapter4-payload-handoff-format.html#node-pci-r

Reply:
please remove node name check as it can be any naming.

Reply:
resolved.